### PR TITLE
Improve `useOnyx` performance

### PR DESCRIPTION
**Pull request by fabioh8010**
### Details

@hurali97 

This PR improves `useOnyx` performance by introducing two changes:

1. Optimises `getSnapshot()` function to only get the newest cached value from Onyx IF we are connecting to `Onyx.connect()` for the first time OR after the `Onyx.connect()` callback is fired, since `getSnapshot()` can be called several times by React and we only want to update the cache under this specific condition.
2. Optimises previous cached value / new cached value comparison to only use `deepEquals` IF `key` is a collection key AND `selector` is defined. `shallowEquals` is enough for the other cases and will be used instead.

### Related Issues

https://github.com/Expensify/App/issues/42421

### Manual Tests

Some common tasks on E/App:

1. Start a new chat and search for someone (uses `useOnyx`).
2. Send a message to someone.
3. Open someone's profile and change settings (uses `useOnyx`).
4. Open Preferences and change settings (uses `useOnyx`).

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/2f95b10b-82cd-444f-9c3e-5601b3f7b1e8

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/7e9dc885-f706-4965-9673-e1bc915e7b35

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/08b8e24d-978e-4d15-b49b-23ee1f1302f7

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/41fbe307-c83f-4d61-abee-69cdab222364

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/dbb60761-598b-4f44-87ac-607897bbff65

https://github.com/Expensify/react-native-onyx/assets/20051562/6caecf22-2d74-400a-9407-0f4eaf2ceab2

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/react-native-onyx/assets/20051562/8ee94825-4bf7-4c2c-84df-6f19d8416cbc

</details>

### CODE REVIEW

File: lib/useOnyx.ts [hunk 1]
```diff
@@ -139,24 +153,35 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = UseOnyxValue<TKey>>(key: T
...
+        if (isFirstConnectionRef.current && newValueRef.current === undefined && options?.initialValue !== undefined) {
+            newValueRef.current = options?.initialValue as CachedValue<TKey, TReturnValue>;
             newFetchStatus = 'loaded';
         }
 
+        // We do a deep equality check if we are subscribed to a collection key and `selector` is defined,
+        // since each `OnyxUtils.tryGetCachedValue()` call will generate a plain new collection object with new records as well,
+        // all of them created using the `selector` function.
+        // For the other cases we will only deal with object reference checks, so just a shallow equality check is enough.
+        let areValuesEqual = false;
```
Comment (janicduplessis) [inline lib/useOnyx.ts:171]: Nit: no need to initialize here.

```suggestion
        let areValuesEqual: boolean;
```

File: tests/unit/useOnyxTest.ts [hunk 2]
```diff
@@ -195,7 +195,27 @@ describe('useOnyx', () => {
             expect(result.current[1].status).toEqual('loaded');
         });
 
-        it('should not change selected data if a property outside the selector was changed', async () => {
+        it('should not change selected data if a property outside tha data was changed', async () => {
```
Comment (janicduplessis) [inline tests/unit/useOnyxTest.ts:198]:
```suggestion
        it('should not change selected data if a property outside that data was changed', async () => {
```

**Review by janicduplessis — APPROVED** [review]
Overall looks good to me. Do you have any performance measurements before / after this change? Maybe using the same flow described in the linked issue.

**Comment by fabioh8010** [comment]
@mountiny This PR is still waiting for the Onyx bump, right?

**Comment by mountiny** [comment]
Correct

**Review by chrispader — APPROVED** [review]
LGTM!

File: lib/useOnyx.ts [hunk 3]
```diff
@@ -131,24 +146,37 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey
...
         // If the cache was set for the first time, we also update the cached value and the result.
-        const isCacheSetFirstTime = cachedValueRef.current === null && hasCacheForKey;
-        if (isCacheSetFirstTime || !deepEqual(cachedValueRef.current ?? undefined, newValue)) {
-            cachedValueRef.current = newValue;
-            resultRef.current = [cachedValueRef.current as CachedValue<TKey, TReturnValue>, {status: newFetchStatus ?? 'loaded'}];
+        const isCacheSetFirstTime = previousValueRef.current === null && hasCacheForKey;
+        if (isCacheSetFirstTime || !areValuesEqual) {
+            previousValueRef.current = newValueRef.current;
+
+            // If the new value is `null` we default it to `undefined` to ensure the consumer get a consistent result from the hook.
```
Comment (mountiny) [inline lib/useOnyx.ts:178]:
```suggestion
            // If the new value is `null` we default it to `undefined` to ensure the consumer gets a consistent result from the hook.
```

**Review by mountiny — APPROVED** [review]
Looks good to me, one NAB

**Pull request merged**